sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProtoTest.java

```diff
@@ -0,0 +1,304 @@
+            .collect(
+                Collectors.toMap(FieldDescriptorProto::getName, FieldDescriptorProto::getType));
+    assertEquals(expectedTypes, types);
+  }
+
+  @Test
+  public void testNestedFromTableSchema() {
+    DescriptorProto descriptor =
+        TableRowToStorageApiProto.descriptorSchemaFromTableSchema(NESTED_TABLE_SCHEMA);
+    Map<String, Type> expectedBaseTypes =
+        BASE_TABLE_SCHEMA_PROTO.getFieldList().stream()
+            .collect(
+                Collectors.toMap(FieldDescriptorProto::getName, FieldDescriptorProto::getType));
+
+    Map<String, Type> types =
+        descriptor.getFieldList().stream()
+            .collect(
+                Collectors.toMap(FieldDescriptorProto::getName, FieldDescriptorProto::getType));
+    Map<String, String> typeNames =
+        descriptor.getFieldList().stream()
+            .collect(
+                Collectors.toMap(FieldDescriptorProto::getName, FieldDescriptorProto::getTypeName));
+    assertEquals(2, types.size());
+
+    Map<String, DescriptorProto> nestedTypes =
+        descriptor.getNestedTypeList().stream()
+            .collect(Collectors.toMap(DescriptorProto::getName, Functions.identity()));
+    assertEquals(2, nestedTypes.size());
+    assertEquals(Type.TYPE_MESSAGE, types.get("nestedvalue1"));
+    String nestedTypeName1 = typeNames.get("nestedvalue1");
+    Map<String, Type> nestedTypes1 =
+        nestedTypes.get(nestedTypeName1).getFieldList().stream()
+            .collect(
+                Collectors.toMap(FieldDescriptorProto::getName, FieldDescriptorProto::getType));
+    assertEquals(expectedBaseTypes, nestedTypes1);
+
+    assertEquals(Type.TYPE_MESSAGE, types.get("nestedvalue2"));
+    String nestedTypeName2 = typeNames.get("nestedvalue2");
+    Map<String, Type> nestedTypes2 =
+        nestedTypes.get(nestedTypeName2).getFieldList().stream()
+            .collect(
+                Collectors.toMap(FieldDescriptorProto::getName, FieldDescriptorProto::getType));
+    assertEquals(expectedBaseTypes, nestedTypes2);
+  }
+
+  @Test
+  public void testRepeatedDescriptorFromTableSchema() {
+    TableRowToStorageApiProto.descriptorSchemaFromTableSchema(BASE_TABLE_SCHEMA);
+  }
+
+  private static final TableRow BASE_TABLE_ROW =
+      new TableRow()
+          .set("stringValue", "string")
+          .set(
+              "bytesValue", BaseEncoding.base64().encode("string".getBytes(StandardCharsets.UTF_8)))
+          .set("int64Value", "42")
+          .set("intValue", "43")
+          .set("float64Value", "2.8168")
+          .set("floatValue", "2.817")
+          .set("boolValue", "true")
+          .set("booleanValue", "true")
+          .set("timestampValue", "43")
+          .set("timeValue", "00:52:07[.123]|[.123456] UTC")
+          .set("datetimeValue", "2019-08-16 00:52:07[.123]|[.123456] UTC")
+          .set("dateValue", "2019-08-16")
+          .set("numericValue", "23.4");
+
+  private void assertBaseRecord(DynamicMessage msg) {
```

Comment:
Is this just comparing to the code itself? Could we have static expectations?

Reply:
done. fixed both tests to have static expectations.